Handreiking API Design/Handreiking APi Design.md

```diff
@@ -54,14 +63,21 @@ De naamgevingsconventie is nl. als volgt:
 
 Voorbeelden hiervan zijn:
 
-* `ZaakContactmoment`
-* `VerzoekInformatieobject`
-* `ZaakBesluit`
-* `BesluitInformatieobject`
+* `zaakverzoek`
+* `zaakcontactmoment`
+* `verzoekinformatieobject`
+* `zaakbesluit`
+* `besluitinformatieobject`
 
 In een gegevensmodel kunnen in dit geval 2 klasses opgenomen zijn met dezelfde
 naam maar elk opgenomen in een ander domein.
 
+>Deze naamgevingsconventie is nog aan discussie onderhevig. Tijdens de webinar voor de ZGW API's van 16 december 2019 heeft Henri Korver een ander voorstel gedaan waarbij de naamgeving de volgende conventie volgt:
+>
+>`[naam van de gerelateerde class in het source domein][naam van de gerelateerde class in het target domein]`.
+>
+>Indien besloten wordt deze conventie te gaan volgen is er natuurlijk geen sprake meer van dat in een gegevensmodel 2 klasses opgenomen zijn met dezelfde naam.
+
```

Comment:
Het lastige van dit voorstel is dat de API aangepast moet worden wanneer vanuit een nieuwe/andere API (register) naar het bestaande object verwezen wordt. 

Voorbeeld: Er komt een Boodschappenbriefjes API waarbij een Boodschappenbriefje vastgelegd kan worden als InformatieObject.  In de huidige situatie hoeft er niets te veranderen aan de DRC API waar de InformatieObject resource leeft, er kan immers verwezen worden naar een ObjectInformatieObject waar nu een verwijzing naar een Boodschappenbriefje in opgenomen kan worden.

Wanneer er expliciete relatieklassen gemodelleerd worden en de DRC API dus een nieuwe relatieklasse BoodschappenbriefjeInformatieObject moet gaan ondersteunen moet er een nieuwe versie van de DRC API uitgebracht worden die feitelijk niets wezenlijks toevoegt. Het brengt echter wel de verplichting voor DRC leveranciers om de nieuwe API in te bouwen, ongeacht of er iets met een Boodschappenbriefje API gedaan wordt.

Tevens vervuilt het expliciet maken van de niet-dominante relatieklasses het gegevens model met relatieklasses die helemaal niet altijd van toepassing hoeven te zijn. Om de API correct te ondersteunen moet een leverancier dit echter wel inbouwen. 

Ik zou daarom de voorkeur er aan geven in het dominante domein de relatieklasses sterk getypeerd te houden en in het niet-dominante domein de generieke ObjectKlasse relatieklasses aan te houden.

Reply:
Ik vermoed dat het principe dat je hierboven beschrijft ook op de naamgevingsconventie van Henri los kunt laten. Het voorstel van Henri wijzigt de volgorde van de onderdelen in de naamgeving. Je kunt aan de conventie echter nog steeds toevoegen dat als er potentieel meerdere target domeinen zijn je de naam 'object' gebruikt. De relatie-entiteit 'objectverzoek' in het klantinteracties domein zou dan hernoemd moeten worden naar 'verzoekobject'.

Reply:
Op zich moet de API dan bij een toevoeging aangepast worden, maar de software van de leverancier moet dan sowieso aan die nieuwe functionaliteit worden aangepast. Dan is de API aanpassen niet het grootste probleem lijkt mij. Aangezien het een toevoeging is zal dit geen backward incompatible wijziging zijn. Ergo, Als de leverancier de nieuwe functionaliteit wil gaan toepassen moet de software sowieso aangepast worden en is een API aanpassen daar onderdeel van en als hij die nieuwe functionaliteit niet wil aanpassen imlementeert hij deze nieuw minor versie niet.

Reply:
Hierboven worden volgens mij 2 verschillende discussies gevoerd.

1. Michiel en ik discussiëren er over of het toepassen van de 'xxxxxxObject' notatie ook op het voorstel van Henri van toepassing kan zijn. Volgens mij is dat het geval. Het gebruik van de string 'Object' in de naam van de relatieklasse wordt in deze discussie echter niet ter discussie gesteld.

2. Johan stelt het gebruik van de string 'Object' in de naam van de relatieklasse wel ter discussie. Hij geeft aan dat hij liever specifiek benoemde relatieklasses gebruikt ziet worden. @JohanBoer: als ik het mis heb dan hoor ik dat graag.

Het lijkt me goed om hier eerst over te discussiëren voordat ik aanpassingen aanbreng aan het onderliggende document.
@joeribekker, @sergei-maertens, graag jullie commentaar.

Reply:
Ik kan ook beide uitgangspunten noteren op de wijze waarop ik dat eerder heb gedaan en de discussie doorschuiven naar een moment waarop we gezamenlijk over API Design gaan discussiëren.

Reply:
> 
> 
> Op zich moet de API dan bij een toevoeging aangepast worden, maar de software van de leverancier moet dan sowieso aan die nieuwe functionaliteit worden aangepast. Dan is de API aanpassen niet het grootste probleem lijkt mij. Aangezien het een toevoeging is zal dit geen backward incompatible wijziging zijn. Ergo, Als de leverancier de nieuwe functionaliteit wil gaan toepassen moet de software sowieso aangepast worden en is een API aanpassen daar onderdeel van en als hij die nieuwe functionaliteit niet wil aanpassen imlementeert hij deze nieuw minor versie niet.

Dit gaat over koppeltabellen tussen domeinen/API's. Bijvoorbeeld de koppeling tussen zaak en informatieobject.  In dit geval wordt de relatie gelegd vanuit de zaak en is er sprake van een ZaakInformatieobject koppeling in de ZRC. In de DRC hoeft alleen vastgelegd te worden dat het informatieobject gekoppeld is aan "iets". Dit om te voorkomen dat in de DRC een gegeven verwijderd wordt en daarmee een relatie in een ander domein (waar de DRC geen controle over heeft) ongeldig wordt.
Het doel van de relatie ObjectInformatieobject is dus om aan te geven dat een Informatieobject gekoppeld is aan iets wat buiten de DRC bestaat en daarom het informatieobject niet zomaar verwijderd oid mag worden.  Voor het overige kun je aan de hand van de verwijzing (uri naar " iets" ) afleiden wat "iets"  precies is, een zaak, een besluit, een ?.

Wanneer je een ObjectInformatieobject sterk zou moeten typeren betekent dat automatisch dat wanneer Informatieobjecten ook vanuit een ander domein gekoppeld moeten kunnen worden de DRC aangepast moet worden. In het voorbeeld: Wanneer een informatieobject aan een nieuwe Boodschappenbriefje API gekoppeld moet worden (voor de relatie Boodschappenbriefje is vastgelegd als Informatieobject) moet een nieuwe versie van de DRC gemaakt worden terwijl er technisch en functioneel helemaal niets toegevoegd wordt **_aan de functionaliteit van de DRC_**.  Het enige wat in de DRC vastgelegd moet worden is dat een Informatieobject gekoppeld is aan " iets". Dat er in het lijstje "iets" een keuze toegevoegd wordt voegt voor de applicatie DRC zelf helemaal niets toe.

Afzien van implementatie van de nieuwste versie van de API is niet mogelijk, er kan immers altijd een keer een nieuwe versie van de API komen die je wel wilt implementeren en dan moet je ook die onnodige harde koppeltabel implementeren waar je niets mee doet.

Daarnaast maak je het gemeenten nodeloos moeilijk omdat implementatie van een Boodschappenbriefjes API ineens afhankelijk wordt van de implementatie van de nieuwe versie van de DRC API.   Volgens mij wilden we die afhankelijkheid nou juist kwijt.

Tot slot: koppelen van een Informatieobject aan "iets" is geen nieuwe functionaliteit voor de DRC applicatie. Door dit sterk te typeren in de API dwing je DRC leveranciers nieuwe versies uit te brengen omwille van non-redenen. Het voegt niets toe (functioneel of technisch, het is immers een koppeltabel met 2 verwijzingen, verder niets) maar kost wel tijd, geld, energie. Verspilde tijd, geld en energie in mijn ogen. Verspilling waar niemand blij van wordt, gemeenten niet maar ook leveranciers niet.

Reply:
> Wanneer je een ObjectInformatieobject sterk zou moeten typeren betekent dat automatisch dat wanneer Informatieobjecten ook vanuit een ander domein gekoppeld moeten kunnen worden de DRC aangepast moet worden. In het voorbeeld: Wanneer een informatieobject aan een nieuwe Boodschappenbriefje API gekoppeld moet worden (voor de relatie Boodschappenbriefje is vastgelegd als Informatieobject) moet een nieuwe versie van de DRC gemaakt worden terwijl er technisch en functioneel helemaal niets toegevoegd wordt aan de functionaliteit van de DRC. Het enige wat in de DRC vastgelegd moet worden is dat een Informatieobject gekoppeld is aan " iets". Dat er in het lijstje "iets" een keuze toegevoegd wordt voegt voor de applicatie DRC zelf helemaal niets toe.

Michiel, als ik het goed zie wordt juist aan het 'objectinformatieobject' een 'objectType' attribuut toegevoegd met een enumeratie (nu kan dit de waardes 'besluit' en 'zaak' hebben).
Volgens mij betekent dit dat als je een informatieobject ook aan een 'boodschappenbriefje' wil kunnen koppelen je daarvoor een voorziening moet treffen in de 'Boodschappenbriefje API' (het objectType 'boodschappenbriefjeinformatieobject') maar ook in de DRC API. Je betoog zou denk ik wel opgaan als het attribuut 'objectType' juist aan de voorziening in de 'Boodschappenbriefje API' zou toekennen. Daar is het echter juist niet noodzakelijk omdat daar al duidelijk wordt waaraan gekoppeld wordt.

Reply:
Hmm, volgens mij maakt het allemaal niet zo veel uit :)

In beide scenario's (1. gebruik van "object notatie" vs 2. specifieke target naamgeving) moet de API worden bijgewerkt. Immers, bij (1) moet er in de body een `type` (enumeratie) attribuut zijn die aangeeft waar een verwijzing naar wordt gemaakt. Deze enumeratie moet bij een nieuw target worden uitgebreid, en bij (2) komt er een hele nieuwe resource bij. Beide overigens minor-changes!

Optie 1 is net iets makkelijker vermoed ik (but time will tell) omdat je je software in sommige gevallen niet hoeft aan te passen als er een enum-value bij komt. Het gebruik van een nieuwe resource is iets ingrijpender.

Waar ik **WEL** onderscheid in wil maken is de intentie van de relatie. Ik zou zeker niet `VerzoekObject` introduceren als je hier zowel naar een `Zaak` als een `InformatieObject` kan wijzen waarbij in het geval van `Zaak` het betekent "Verzoek leidt tot Zaak" en bij `InformatieObject` "Verzoek heeft als bijlage InformatieObject".

Reply:
@melsk-r Ik zou er nog iets in zetten over het conceptuele onderscheid voordat er wordt opgeschreven dat zodra er relaties zijn naar meer dan 1 object, ze samengevoegd worden onder "<origin>object".

Reply:
Wat bedoel je precies?

Reply:
De vraag die bij mij opkomt is:

Wanneer gebruik ik <master><slave> en wanneer gebruik ik object<slave> / <master>object notaties?

Denk dat er duidelijke voorbeelden moeten komen. Bijvoorbeeld:

In de Klantinteractie API, als ik wil verwijzen van een Verzoek naar een Zaak, waarbij de "master"-relatie in de Zaken API ligt, dan wordt het ZAAK-VERZOEKen. Stel ik wil nu ook van Verzoek (in Klantinteractie API) naar InformatieObject (in Documenten API) wijzen waarbij de "master"-relatie ligt in de Documenten API. Moet ZAAK-VERZOEKen dan worden omgedoopt tot OBJECT-VERZOEKen?

Mijn opmerking gaat erover dat het antwoord afhankelijk is van het type relatie (of niet?). Wanneer voeg je relaties samen onder "OBJECT" en wanneer laat je ze gescheiden, zodat er ZAAK-VERZOEKen en INFORMATIEOBJECT-VERZOEKen ontstaat?